### PR TITLE
Fix AOTAutograd unused output backward pruning

### LINKED ISSUE #131794: [torch.compile] Unused differentiable outputs lead to unnecessary gradient computation when compared to eager
The following code:
```py
import torch

@torch.compile(backend="aot_eager")
def f(xs): # E: Function is missing a type annotation  [no-untyped-def]
    result = []
    for x in xs:
        result.append(x.sin())
    return result

xs = [torch.randn([], requires_grad=True) for _ in range(3)]
ys = f(xs)
ys[0].backward()
```
produces the following backward graph:
```
def forward(self, primals_1: "f32[][]cpu", primals_2: "f32[][]cpu", primals_3: "f32[][]cpu", tangents_1: "f32[][]cpu", tangents_2: "f32[][]cpu", tangents_3: "f32[][]cpu"):
     # File: /home/rzou/dev/debug-cpu0/pt-debug-cpu0/foo.py:7 in f, code: result.append(x.sin())
    cos: "f32[][]cpu" = torch.ops.aten.cos.default(primals_3);  primals_3 = None
    mul: "f32[][]cpu" = torch.ops.aten.mul.Tensor(tangents_3, cos);  tangents_3 = cos = None
    cos_1: "f32[][]cpu" = torch.ops.aten.cos.default(primals_2);  primals_2 = None
    mul_1: "f32[][]cpu" = torch.ops.aten.mul.Tensor(tangents_2, cos_1);  tangents_2 = cos_1 = None
    cos_2: "f32[][]cpu" = torch.ops.aten.cos.default(primals_1);  primals_1 = None
    mul_2: "f32[][]cpu" = torch.ops.aten.mul.Tensor(tangents_1, cos_2);  tangents_1 = cos_2 = None
    return [mul_2, mul_1, mul]
```
We're only differentiating through one output, so there should only be a single call to cos. Instead, AOTAutograd is materializing all of the None gradients as zeros.

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.16.0) (oldest at bottom):
* __->__ #186355

AOTAutograd's custom autograd.Function let autograd materialize undefined grad outputs as zeros before the compiled backward saw them. A backward through one differentiable output therefore compiled and ran independent unused branches, and unused inputs received zero gradients where eager mode leaves them undefined.

Preserve undefined grad outputs and specialize the already-traced lazy backward for undefined visible user-output tangents. The specialization propagates symbolic zeros through supported linear, view, and backward patterns so FX can remove unused branches. If an operation cannot be simplified, run the unspecialized backward with only its missing tangents materialized, then mask input gradients that depend solely on pruned outputs. This keeps eager `None`-gradient semantics without retracing the joint graph or disturbing saved-tensor and DDP lifetime rules.

Make compiled autograd use the same visible-output specialization key; hidden mutated-input and intermediate-base positions must not change its semantics. Its cached graph closes over only a static prototype schema and reads per-invocation, zero-storage size/stride carriers from the runtime context, so dynamic shapes are not frozen and forward activations are not retained. Dense, sparse, jagged, DTensor, and wrapper-subclass tangents are reconstructed directly in the flat backward ABI, cloning only structural state that is part of the tangent value. The common single-tangent case does not allocate prototype carriers.

Prune specialized boxed argument lists in place so `steal_args` continues to release saved activations on both fresh and cached paths. Correct subclass leaf indexing to account for intervening symbolic metadata arguments.

## BC-breaking?

Yes. Unused independent inputs now keep `.grad` as `None`, matching eager mode, and `torch.autograd.grad(..., allow_unused=False)` can raise where the old compiled behavior returned zero. The behavior is documented, and setting `torch._functorch.config.aot_autograd_prune_unused_outputs=False` before the compiled forward restores legacy zero materialization.

The implementation specializes the existing backward graph instead of retracing the joint graph for each tangent pattern. This is more local, preserves existing backward compiler/cache behavior, and retains a correctness fallback for unsupported zero propagation.

Fixes #131794
Generated by my agent

Test Plan:
- `python test/functorch/test_aotdispatch.py -k unused_differentiable` (27 passed)
- `python test/functorch/test_aotdispatch.py -k grad_output_prototypes` (16 passed)
- `python test/functorch/test_aotdispatch.py -k compiled_autograd` (7 passed)
- `python test/dynamo/test_backward_higher_order_ops.py` (7 passed)
- `python test/inductor/test_compiled_autograd.py TestCompiledAutograd.test_dtensor_unused_output_fallback TestCompiledAutograd.test_tensor_subclass_basic` (2 passed)
- `lintrunner -a docs/source/user_guide/torch_compiler/torch.compiler_backward.md test/dynamo/test_backward_higher_order_ops.py test/functorch/test_aotdispatch.py test/inductor/test_compiled_autograd.py torch/_dynamo/compiled_autograd.py torch/_functorch/_aot_autograd/runtime_wrappers.py torch/_functorch/config.py`
- `git diff --check`

cc @ezyang @gchanan @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @azahed98 @xmfan @aditvenk

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I updated the PR to address the CI failures in compiled autograd and AOTAutograd subclass tangent handling. The new patch passes the updated backward prologue metadata through compiled autograd, masks copied AOT backward outputs that depend only on undefined grad outputs, avoids retaining ordinary output storage for zero-grad materialization, and adds a compiled-autograd unused-output regression.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I updated the PR to address the reported CI failures and review cleanup. The compiled-autograd path now specializes the copied AOT backward graph for undefined grad outputs instead of materializing concrete zero tangents, which fixes the dynamic-shape guard failure in TestCompiledAutograd.test_mismatch_fake_tensor_mode_dynamic_shape. The AOTAutograd tangent accounting now excludes ordered-effect token tangents, which fixes TestAOTModuleSimplified.test_grads_no_force_contiguous_subclass. I also removed the unused zero-simplification sentinel and replaced the placeholder list lookup with an index map.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I updated the PR to address the remaining compiled-autograd CI failure in TestCompiledAutograd.test_verbose_logs_aot_dispatcher_nodes_hop. The failure was a stale verbose-log expectation: the new AOT backward pruning correctly removes the unused tangent and add branch, so the expected node list now matches the optimized graph.

File: torch/_dynamo/compiled_autograd.py [hunk 1]
```diff
@@ -501,6 +505,43 @@ def proxy_call_aot_backward(
         bw_prologue_fn = CompiledFunction._bw_prologue_fn
         bw_epilogue_fn = CompiledFunction._bw_epilogue_fn
         del CompiledFunction
+        undefined_grad_out_indices = tuple(
+            i for i, grad in enumerate(inputs) if grad is None
+        )
+        if undefined_grad_out_indices:
+            undefined_grad_out_indices_set = set(undefined_grad_out_indices)
+            grad_output_prototypes = tuple(
```
Comment (bobrenjc93) [inline torch/_dynamo/compiled_autograd.py]: `grad_output_prototypes` is closed over by the `allow_in_graph`'d `call_aot_bwd_prologue` and therefore baked into the cached compiled-autograd graph. This directly violates the "NOTE: we should only close over constants" invariant 14 lines above: `_grad_output_prototype` returns `x.detach()` for wrapper subclasses / non-strided tensors (a real tensor from *this* iteration, kept alive for the lifetime of the cached CA graph = an activation leak), and returns `_GradOutputPrototype` with a concrete `torch.Size`/stride for plain tensors, which is wrong the moment the CA graph is reused with dynamic shapes (the fallback path would materialize zero tangents with iteration-1 shapes). The prototypes need to be derived from the runtime `ctx` inside the proxied prologue call, not captured at trace time.

File: torch/_dynamo/compiled_autograd.py [hunk 2]
```diff
@@ -501,6 +505,43 @@ def proxy_call_aot_backward(
...
+                prototype if i in undefined_grad_out_indices_set else None
+                for i, prototype in enumerate(ctx._aot_grad_output_prototypes)
+            )
+        else:
+            grad_output_prototypes = ()
+        pruned_output_indices = (
```
Comment (bobrenjc93) [inline torch/_dynamo/compiled_autograd.py:545]: CA passes the raw `undefined_grad_out_indices` here, but the eager path (`_AutogradBackwardCompiler.get_pruned_backward_output_indices`) first narrows to `_specializable_user_grad_output_indices(...)` and returns `()` when that is empty. So when the undefined tangents sit at mutated-input or intermediate-base grad-output positions, compiled autograd masks input grads to `None` while eager AOTAutograd returns zeros for the same model. Compiled autograd is supposed to be semantics-preserving w.r.t. the eager AOT backward; please use the same narrowing (and add a test that exercises a mutated-input grad output under CA).

File: torch/_functorch/_aot_autograd/runtime_wrappers.py [hunk 3]
```diff
@@ -2566,6 +2568,615 @@ class AOTDispatchAutogradCompileSpec:
...
+        x = x.alias
+    if isinstance(x, torch.Tensor):
+        if not is_traceable_wrapper_subclass(x) and x.layout is torch.strided:
+            return _GradOutputPrototype(
+                x.size(),
+                x.stride(),
+                x.dtype,
+                x.device,
+            )
+        return x.detach()
```
Comment (bobrenjc93) [inline torch/_functorch/_aot_autograd/runtime_wrappers.py]: `x.detach()` on a traceable wrapper subclass (DTensor, TwoTensor, NJT) or a sparse output shares storage with the forward output, and this is stashed on `ctx` on *every* compiled forward regardless of whether any grad ends up undefined. That pins the forward output's memory until backward runs, which is a peak-memory regression for exactly the DTensor/torchtitan workloads this PR is CI-labelled for. Capture layout metadata only (extend `_GradOutputPrototype` with the subclass type + inner metas, or defer to `fw_metadata.traced_tangents`), or only capture when `subclass_tangent_meta` says it's needed.

File: torch/_functorch/_aot_autograd/runtime_wrappers.py [hunk 4]
```diff
@@ -3526,10 +4282,19 @@ def _backward_impl(ctx: Any, all_args: list[Any]) -> Any:
                 saved_tensors_use_once = (
                     not torch._C._autograd._get_current_graph_task_keep_graph()
                 )
-                compiled_bw = backward_compiler.get_or_compile(
-                    saved_tensors_use_once=saved_tensors_use_once
+                undefined_grad_out_indices = getattr(
+                    ctx, "_undefined_grad_out_indices", ()
+                )
+                compiled_bw, all_args, pruned_output_indices = (
```
Comment (bobrenjc93) [inline torch/_functorch/_aot_autograd/runtime_wrappers.py:4758]: Rebinding `all_args` to a new pruned list defeats `steal_args=True`. The codegen'd `_compiled_backward` frame still holds the original list (`all_args = _prologue_(...)`; `_impl_(_ctx_, all_args)`), so the saved activations in it are no longer freed incrementally as the compiled backward consumes its inputs - they stay alive for the entire backward. For a PR whose motivation is efficiency, this is a significant memory regression on precisely the specialized path. Prune in place (e.g. build the kept list and then `all_args.clear()` on the original) instead of returning a fresh list.

File: torch/_functorch/_aot_autograd/runtime_wrappers.py [hunk 5]
```diff
@@ -2704,7 +3315,14 @@ def finalize(self, ctx: Any, fw_outs: Sequence[Any]) -> tuple[Any, ...]:
...
             if isinstance(x, torch.Tensor) and not raw_returns_meta[i].requires_grad
         ]
+        ctx._aot_grad_output_prototypes = _grad_output_prototypes(
+            raw_returns,
+            num_mutated_runtime_inps
+            + num_outputs
+            + self.metadata.num_intermediate_bases,
+        )
         ctx.mark_non_differentiable(*fw_outs_not_requiring_grad)
+        ctx.set_materialize_grads(False)
```
Comment (bobrenjc93) [inline torch/_functorch/_aot_autograd/runtime_wrappers.py]: This is a user-visible semantics change (unused independent inputs now get `.grad is None` instead of zeros, and `torch.autograd.grad(..., allow_unused=False)` will now raise where it previously succeeded), yet the PR is labelled "topic: not user facing" and there is no `torch._functorch.config` flag to turn any of this off. Given the blast radius - every compiled training backward, plus a new heuristic FX rewrite and an extra backward compile - please add a config kill switch (defaulted on if you like) and a BC note, so this can be bisected/disabled in production without a revert.

File: torch/_functorch/_aot_autograd/runtime_wrappers.py (hunk 3 above)
Comment (jansel) [reply torch/_functorch/_aot_autograd/runtime_wrappers.py]: My agent says detached activation prototypes have been removed. Dense, sparse, jagged, DTensor, and wrapper tangents now use metadata-only schemas plus runtime zero-storage carriers, with wrapper tangents materialized directly as flat ABI leaves and only jagged structural state cloned. Storage-expiry and subclass/layout regressions cover this.

File: torch/_dynamo/compiled_autograd.py (hunk 2 above)
Comment (jansel) [reply torch/_dynamo/compiled_autograd.py:545]: My agent says compiled autograd now derives and uses the same visible-user-output specialization key as normal AOTAutograd. A mutated-input regression verifies that hidden grad-output positions still produce the eager-equivalent zero gradient.

File: torch/_dynamo/compiled_autograd.py (hunk 1 above)
Comment (jansel) [reply torch/_dynamo/compiled_autograd.py]: My agent says this is addressed in the latest update: compiled autograd now closes over only a static prototype schema and fetches per-invocation, zero-storage metadata carriers from ctx through an explicit FX getattr. A dynamic-size 4/7 cache-reuse regression covers the fallback path.

File: torch/_functorch/_aot_autograd/runtime_wrappers.py (hunk 4 above)
Comment (jansel) [reply torch/_functorch/_aot_autograd/runtime_wrappers.py:4758]: My agent says both the fresh-specialization and cache-hit paths now prune the original all_args list in place. A boxed backward regression checks that saved storage is released on both calls.

File: torch/_functorch/_aot_autograd/runtime_wrappers.py (hunk 5 above)
Comment (jansel) [reply torch/_functorch/_aot_autograd/runtime_wrappers.py]: My agent says a default-on torch._functorch.config.aot_autograd_prune_unused_outputs kill switch now restores legacy zero materialization when disabled. The PR and compiler-backward documentation include the BC impact, and the labels now identify the BC break and AOTAutograd release-note category.

**Comment by jansel** [comment]
My agent says I rebased onto current main and addressed all five review findings: runtime-safe dynamic prototypes without forward-activation retention, visible-output narrowing in compiled autograd, dynamic subclass leaf accounting, in-place boxed argument pruning, and a kill switch with BC documentation. The focused AOTAutograd, compiled-autograd, DTensor/subclass, prototype/storage, higher-order, and lint suites in the updated Test Plan all pass.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says this PR is being closed as part of a cleanup of open agentic PRs to focus review bandwidth on a smaller set; the linked issue remains open.
